### ISSUE TITLE
Generic collection fixtures

### ISSUE THREAD

**Issue by lonix1** [xunit/xunit#2557]
### Discussed in https://github.com/xunit/xunit/discussions/2425

<div type='discussions-op-text'>

<sup>Originally posted by **lonix1** November 24, 2021</sup>
@bradwilson I didn't know whether to file this as an issue or discussion, sorry!

A generic fixture and support classes:

````cs
public class Fixture<T> : IAsyncLifetime, IDisposable {
  public Fixture() { }
  public void Dispose() { }
  public Task InitializeAsync() => Task.CompletedTask;
  public Task DisposeAsync() => Task.CompletedTask;
}


[CollectionDefinition("collection")]
public class FixtureCollection<T> : ICollectionFixture<Fixture<T>> { }


[Collection("collection")]
public abstract class TestBase<T> {
  protected TestBase(Fixture<T> fixture) => _fixture = fixture;
  protected readonly Fixture<T> _fixture;
}
````

I stripped down those classes to show a minimal sample (but there is "stuff" in them).

And the test class:
````cs
public class Foo { }


public class Tests : TestBase<Foo> {
  public Tests(Fixture<Foo> fixture) : base(fixture) { }
  [Fact] public void Test1() { }
  [Fact] public void Test2() { }
}
````

Tests fail:
> The following constructor parameters did not have matching fixture data: Fixture`1 fixture

If I make this change then it works, but I can't use it as a collection fixture:
````cs
//[Collection("collection")]                                      // removed
public abstract class TestBase<T> : IClassFixture<Fixture<T>> {   // added IClassFixture
````

I read [the docs](https://xunit.net/docs/shared-context).

**Is this use case supported?**</div>

**Comment by FLAMESpl** [comment]
Hi @bradwilson this issue is half a year old, can we get any comment on this?
My example is also not working altough everything looks according to the book:

```csharp
using Xunit;

namespace XUnitFixtureIssue;

[CollectionDefinition(NAME)]
public class MyCollection : ICollectionFixture<MyCollectionFixture>
{
    public const string NAME = nameof(MyCollection);
}

public class MyCollectionFixture
{
}

public abstract class MyClassFixtureBase
{
    protected MyClassFixtureBase(MyCollectionFixture fixture)
    {
    }
}

public class MyDerivedClassFixture : MyClassFixtureBase
{
    public MyDerivedClassFixture(MyCollectionFixture fixture) : base(fixture)
    {
    }
}

public abstract class MyTestBase<TFixture> : IClassFixture<TFixture> where TFixture : MyClassFixtureBase
{
    protected MyTestBase(MyClassFixtureBase fixture)
    {
    }

    [Fact]
    public void TestMethod() { }
}

public class MyTest : MyTestBase<MyDerivedClassFixture>
{
    public MyTest(MyDerivedClassFixture fixture) : base(fixture)
    {
    }
}
```

Error message is

```
System.AggregateException: One or more errors occurred. (Class fixture type 'XUnitFixtureIssue.MyDerivedClassFixture' had one or more unresolved const...

System.AggregateException
One or more errors occurred. (Class fixture type 'XUnitFixtureIssue.MyDerivedClassFixture' had one or more unresolved constructor arguments: MyCollectionFixture fixture) (The following constructor parameters did not have matching fixture data: MyDerivedClassFixture fixture)
  Exception doesn't have a stacktrace

Xunit.Sdk.TestClassException
Class fixture type 'XUnitFixtureIssue.MyDerivedClassFixture' had one or more unresolved constructor arguments: MyCollectionFixture fixture
  Exception doesn't have a stacktrace

Xunit.Sdk.TestClassException
The following constructor parameters did not have matching fixture data: MyDerivedClassFixture fixture
  Exception doesn't have a stacktrace

```

**Comment by ultrabstrong** [comment]
I would love to see this implemented for a similar use case:
```c#
public class LoggerFixture<T> : IDisposable
{
    public ILogger<T> Logger { get; } // Microsoft logger
    public LoggerFixture()
    {
        var logger = new LoggerConfiguration().WriteTo.Console().CreateLogger();
        var factory = LoggerFactory.Create(builder => builder.AddSerilog(logger));
        Logger = factory.CreateLogger<T>();
    }
}
``` 
```c#
[CollectionDefinition("LoggerCollection")]
public class LoggerCollection<T> : ICollectionFixture<LoggerFixture<T>> { }
``` 
```c#
 [Collection("LoggerCollection")]
 public class TestClass(LoggerFixture<SutClass> loggerFixture)
 {
     private readonly ILogger<SutClass> _logger = loggerFixture.Logger;
 }
```

**Comment by JamesTerwilliger** [comment]
I've looked into this one and will give it a try in the next few weeks.

**Comment by bradwilson** [comment]
Available in v3 `0.1.1-pre.366` https://xunit.net/docs/using-ci-builds

**Comment by JamesTerwilliger** [comment]
@bradwilson Analyzer fix ready in https://github.com/xunit/xunit.analyzers/pull/180
